helpers.sh

```diff
@@ -137,7 +137,12 @@ function setup_port_forwarding() {
     echo
     echo "== Setting up port forwarding =="
     sleep 5
-    echo "ssh -L $PORT:localhost:$PORT ${RESOURCE} ssh -L $PORT:localhost:$PORT -N $MACHINE &"
-    ssh -L $PORT:localhost:$PORT ${RESOURCE} ssh -L $PORT:localhost:$PORT -N "$MACHINE" &
-
+    if $SSHBOOL
```

Comment:
okay this is taking what I said literally, which is not what I meant! Oy vey, I'll just figure out the name. The `-L` refers to local port forwarding, and `-K` looks like:

```
-K      Enables GSSAPI-based authentication and forwarding (delegation) of GSSAPI credentials to the server.
``` 
so let's assume that GSAPPI works for sherlock and not farmshare, and we can call the variable `USE_GSSAPI`. @akkornel do you know what the difference between these two commands is?

```bash
# works on sherlock
ssh -L $PORT:localhost:$PORT ${RESOURCE} ssh -L $PORT:localhost:$PORT -N "$MACHINE" &

# works for farmshare (not for sherlock?)
ssh "$DOMAINNAME" -l $FORWARD_USERNAME -K -L  $PORT:$MACHINE:$PORT -N  &
```
We basically what a good way to describe the differences between the two to decide on the boolean. Thanks for your help!

Reply:
Hi @vsoch, I hope you're doing well!

I haven't used `-K` before, so I had to look it up to make sure I understood it.

Putting -K on the command line is the same as setting these two options in `ssh_config` (or `~/.ssh/config`):

```
GSSAPIAuthentication yes
GSSAPIDelegateCredentials yes
```

It tells the SSH client to attempt using GSSAPI authentication for the connection to the remote host.  This option is safe to use on both FarmShare and Sherlock!  I was concerned that the `-K` option might be so new as to not be widely supported, but a `git blame` on https://github.com/openssh/openssh-portable/blob/master/ssh.c#L841 shows that it's been implemented for a long time, so by now it's safe to use anywhere.

As for those two commands, I'm not sure at first glance.  Could you tell me the definitions of `$RESOURCE`, `$MACHINE`, and `DOMAINNAME`?

Reply:
I am sorry if I didn't explain this earlier - the reason why the command on farmshare does not work because the compute node has different permissions to log in to. 

The difference between the lines is that we can 

Aka on sherlock - we can pass on our persisting logon shell that we initially bind to - through ssh permissions and still be able to log on to the compute node because the first line connects us to sherlock, and then the second ssh connects us to the compute node because sherlock has permissions setup to ssh into the compute node without reauthenticating. It turns out that on sherlock we can connect to the compute node by passing on our credentials and not having two factor authenticate through control persist that already exists.

Hence, why the double ssh tunnel line works on sherlock because it is set up to pass on our credentials to the compute node if we once login via the login node.   

So when we try to do the same with farmshare - we can authenticate to the login node without duo factor due to the existing tunnel shell. But when we try to establish a secondary tunnel on rice to the compute node, we fail because there is no permissions setup to pass on to that node. Aka that node was never logged on because that node resides on a different place, hence we cannot tunnel into the node without logging into it first and duo-authenticating.

In contrast, farmshare, I have to authenticate every-time I log into a node. So when the double tunnel runs,  it leads to multiple authentication issues which are not resolved because are not properly authenticating ourselves on the compute node on farmshare. Hence, the double tunnel does not work. 

Now Im not sure why sherlock doesnt respond well to the ssh tunnel from domain name to tunnel the port binding. It seems to fail more than anything with connection refused. Because in theory we should be able to pass on our login credentials to the compute node and get away with it - instead of double tunneling first into the login and then to the compute.

Reply:
@akkornel 
$MACHINE is the assigned compute node from either sherlock or farmshare
$RESOURCE is the HPC cluster
$DOMAINNAME is the login node id - either login.sherlock.stanford.edu or rice.stanford.edu

Reply:
@sohams-MAAS 

> Now Im not sure why sherlock doesnt respond well to ssh -K option tunnel the ssh tunnel from domain name to tunnel the port binding. It seems to fail more than anything with connection refused. Because in theory we should be able to pass on our login credentials to the compute node and get away with it - instead of double tunneling first into the login and then to the compute.

The reason why GSSAPI doesn't work on Sherlock compute nodes is because they do not have GSSAPI support.  The Sherlock compute nodes do not support GSSAPI is because GSSAPI requires that the server appear in DNS.  Sherlock compute nodes are on an isolated network, with no direct connectivity to the outside world.  As such, the Sherlock compute nodes do not appear in Stanford (or public) DNS.

That is a big difference between Sherlock and FarmShare: FarmShare compute nodes (the `wheat` and `oat` nodes) do have outside connectivity, and do appear in Stanford/public DNS, so they have support for GSSAPI authentication.  But two-step is still required when connecting to them, because of the [MinSec for Servers](https://uit.stanford.edu/guide/securitystandards#security-standards-servers) requirement for Two-Factor Authentication.

Reply:
I see - so we are back to the original problem of whether we have compute nodes that are isolated from the outside world. I think that can be a proper boolean @vsoch. If nodes are isolated, than we use the double tunnel method of sherlock. 

If nodes are not isolated, then we can use the -K method.